### PR TITLE
Add support to multiple networks

### PR DESCRIPTION
**Pull request by evandrosaturnino**
## Description

Added multi-network support, including updated environment variables, network-specific explorer links, refactored contract integrations and connector handling, and enhanced local storage and minting flow for multi-network compatibility. Fixed ledger and subgraph issues, and centralized network logic.

- [x] - [Update env variables](https://github.com/threshold-network/token-dashboard/pull/785/commits/3db541e778f13120dc77f874cd850bbf42275364)
- [x] - [Update tbtc sdk version in package json](https://github.com/threshold-network/token-dashboard/pull/785/commits/3cf9bc206f54860b8a9ea6bfffbadd92cc79d1c7)
- [x] - [Update explorer links based on network](https://github.com/threshold-network/token-dashboard/pull/785/commits/4b05db181916fc8b7453a7199d78d7e802e3d2f1)
- [x] - [Fix ledger support](https://github.com/threshold-network/token-dashboard/pull/785/commits/ae9a326d75606c7ec37ec81340d46efaec2af8ec)
- [x] - [Add networks centralized folder](https://github.com/threshold-network/token-dashboard/pull/785/commits/10af473cf4bc7e1d51e78f14124570652c363ef5)
- [x] - [Update contract initialization handling](https://github.com/threshold-network/token-dashboard/pull/785/commits/40383bd81bf209f5ddaafedafd2a3049885e3700)
- [x] - [Refactor connector logic to handle multi network](https://github.com/threshold-network/token-dashboard/pull/785/commits/5edbf2623f2bf90b2c55b8bd0ab22bb119a4d7a4)
- [x] - [Add switch support to multiple networks](https://github.com/threshold-network/token-dashboard/pull/785/commits/e24ceb688df5773a9d055dcc6a263389b56294a6)
- [x] - [Fix subgraph issue by migrating to studio services](https://github.com/threshold-network/token-dashboard/pull/785/commits/d0dda9ddbecee799995c957081fae97d59f0b568)
- [x] - [Fix cooldown period for different apps](https://github.com/threshold-network/token-dashboard/pull/785/commits/30ba8c22c31ddd01d1f7b386a63460793bd8d6b7)
- [x] - [Update threshold class to accommodate multiple networks](https://github.com/threshold-network/token-dashboard/pull/785/commits/1dc44894d7e4e3ba66c6ca366ee8645320c888ea)
- [x] - [Update local storage logic handling](https://github.com/threshold-network/token-dashboard/pull/785/commits/a07c8ff6f61bffa7611b249052b0b699e17666a6)
- [x] - [Update minting flow to accommodate multiple networks](https://github.com/threshold-network/token-dashboard/pull/785/commits/c7b1237b85b8c9b0dad32881dffa8e1c426d5896)

## Type of change

- [x]  💅 New Feature (Non-breaking Change)
- [ ] 🐛 Bug fix (Non-breaking Change: Fixes an issue)
- [ ]  🛠️ Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..

## Notice

- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/threshold-network/token-dashboard/pulls) for the same update/change?

## Issue (if applicable)

closes https://github.com/threshold-network/token-dashboard/issues/715
closes https://github.com/threshold-network/token-dashboard/issues/743
closes https://github.com/threshold-network/token-dashboard/issues/750
closes https://github.com/threshold-network/token-dashboard/issues/760
closes https://github.com/threshold-network/token-dashboard/issues/764
closes https://github.com/threshold-network/token-dashboard/issues/787

## Screenshots (if applicable)

image

### CODE REVIEW

**Comment by Shadowfiend** [comment]
Apologies for the late return here—it looks like the preview build isn't loading due to an unset chain id (chain id = 0, which fails everything when trying to fetch the tbtc-v2 artifacts), so it's hard to test the changes here. Any chance we can get that fixed?

Note also that that function (`getArtifact` in `threshold-ts/utils/contract.ts` appears to be set to fail on any network that isn't Ethereum Mainnet or Sepolia). Errors:

Screenshot 2024-11-25 at 10 49 15

**Comment by evandrosaturnino** [comment]
> Apologies for the late return here—it looks like the preview build isn't loading due to an unset chain id (chain id = 0, which fails everything when trying to fetch the tbtc-v2 artifacts), so it's hard to test the changes here. Any chance we can get that fixed?
> 
> Note also that that function (`getArtifact` in `threshold-ts/utils/contract.ts` appears to be set to fail on any network that isn't Ethereum Mainnet or Sepolia). Errors:
> 
> Screenshot 2024-11-25 at 10 49 15

Hey, appreciate the comment, @Shadowfiend - This implementation requires a few new environment variables to be added in the.env file in order for the app to run in default configuration, and there are other to be excluded, as you can see it here: https://github.com/threshold-network/token-dashboard/pull/785/files#diff-9bd30daf6ec3456f77aa9820a3b9d6f37df613de9123d6ed3085a32318b03f7dL89-R99
and
https://github.com/threshold-network/token-dashboard/pull/785/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L132-R141

Basically, the environment variables that have to be added are:

```
REACT_APP_DEFAULT_PROVIDER_CHAIN_ID=1
REACT_APP_ALCHEMY_API=$ALCHEMY_API
REACT_APP_MAINNET_ELECTRUM_PROTOCOL=$MAINNET_ELECTRUM_PROTOCOL
REACT_APP_MAINNET_ELECTRUM_HOST=$MAINNET_ELECTRUM_HOST
REACT_APP_MAINNET_ELECTRUM_PORT=$MAINNET_ELECTRUM_PORT
REACT_APP_TESTNET_ELECTRUM_PROTOCOL=$TESTNET_ELECTRUM_PROTOCOL
REACT_APP_TESTNET_ELECTRUM_HOST=$TESTNET_ELECTRUM_HOST
REACT_APP_TESTNET_ELECTRUM_PORT=$TESTNET_ELECTRUM_PORT
REACT_APP_TBTC_SUBGRAPH_API=$TBTC_SUBGRAPH_API
```

File: src/networks/utils/getRpcUrl.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { EnvVariable } from "../../enums"
+import { getEnvVariable } from "../../utils/getEnvVariable"
+import { SupportedChainIds } from "../enums/networks"
+import { networksAlchemyConfig } from "./networksAlchemyConfig"
+
+const MAIN_ALCHEMY_URL = "g.alchemy.com/v2/"
+
+export const getRpcUrl = (chainId?: number | string) => {
```
Comment (Shadowfiend) [inline src/networks/utils/getRpcUrl.ts:8]: I don't love that we've baked Alchemy in so deep, fwiw.

**Review by Shadowfiend — COMMENTED** [review]
Oi, lost track of this fellow. Can you send me the relevant secrets via Keybase? I'm at https://keybase.io/shadowfiend, you can xref my GitHub proof there.

File: src/networks/utils/getRpcUrl.ts (hunk 1 above)
Comment (evandrosaturnino) [reply src/networks/utils/getRpcUrl.ts:8]: The connection to the RPC URL has become dynamic with the inclusion of support for connecting to different networks. Therefore, the implemented solution, as can be seen below, is that the URL changes dynamically according to the connected network, following the service pattern, for example, if the connected network is Arbitrum on Sepolia, the RPC url would be 'https://arbitrum-sepolia.g.alchemy.com/v2/<SERVICE-API>'. Other services, as Infura, might offer various RPC urls on demand based on the connected networks, but it's expected that the pattern would change.

We can expand the possibilities of different services provided, but some sort of check should be done in this regard. If you think this should be changed, do you have any suggestions for an approach? For example, we could include an environment variable that determines the type of service—e.g., RPC_SERVICE_NAME: 'Alchemy', 'Infura', etc.—and perform validation through an enum providing different URL patterns. Or, alternatively, perhaps a more advanced validation could be done against the API pattern being passed as a parameter in the RPC API in the env file?

https://github.com/threshold-network/token-dashboard/blob/63705c788b4cab1955f8b33351a059ff929d553a/src/networks/utils/getRpcUrl.ts#L8-L17

File: src/hooks/useFetchTvl.ts [hunk 2]
```diff
@@ -62,86 +72,104 @@ export const useFetchTvl = (): [
...
 
-  const fetchOnChainData = useMulticall([
-    {
-      address: multicall?.address!,
-      interface: multicall?.interface!,
+  const calls: Array<ContractCall & { key: string }> = []
+
+  if (keepBonding && multicall) {
+    calls.push({
+      address: multicall.address,
```
Comment (derekpierre) [inline src/hooks/useFetchTvl.ts:79]: Is the postfix `!` no longer needed?

File: src/networks/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    chainId: SupportedChainIds.BaseSepolia,
+    name: "BaseSepolia",
+    layer: Layer.L2,
+    networkType: NetworkType.Testnet,
+    alchemyName: AlchemyName.Base,
+  },
+]
+```
+
+Adding a New Network
```
Comment (derekpierre) [inline src/networks/README.md:126]: Great documentation.

File: src/networks/utils/getRpcUrl.ts (hunk 1 above)
Comment (derekpierre) [reply src/networks/utils/getRpcUrl.ts:8]: I also agree that I don't love how deeply entrenched Alchemy is in the code eg. `alchemyName` in the network definition. Down the road if we decide to switch away from Alchemy, it would be tricky to replace with some other RPC service like Infura or dRPC.

I'm not familiar with the code, so just speculating, but perhaps you could have a generic `API_KEY` and then an additional `API_SERVICE_NAME` variable and a mapping of rpc service name -> rpc config, and the configs have properties/methods to provide/adjust the required names/values/templated strings accordingly. Adding a supported service would involve adding a config entry to the mapping...?

That being said, that can probably be addressed/investigated in a separate PR.

File: src/networks/utils/mappings.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    l2TestNetworksMap[chainId] = name
+  }
+
+  if (layer === Layer.L2 && networkType === NetworkType.Mainnet) {
+    l2MainnetNetworksMap[chainId] = name
+  }
+
+  if (networkType === NetworkType.Testnet) {
+    testNetworksMap[chainId] = name
+  }
```
Comment (derekpierre) [inline src/networks/utils/mappings.ts:34]: Is this code reachable? i.e. is it possible for "layer" to be undefined?

File: src/pages/Staking/HowItWorks/StakingOverview/AuthorizingApplicationsCard.tsx [hunk 5]
```diff
@@ -58,11 +58,11 @@ export const AuthorizingApplicationsCard: FC<ComponentProps<typeof Card>> = (
...
           status={FlowStepStatus.active}
           preTitle="Step 2"
-          title="45 day cooldown"
+          title="period of cooldown"
           size="sm"
         >
-          You must wait a 45 day cooldown to then confirm the deauthorization.
-          This is 1 transaction.
+          There is a deauthorization cooldown period of 45 days for tBTC and
+          Random Beacon apps and 6 months for TACo.
```
Comment (derekpierre) [inline src/pages/Staking/HowItWorks/StakingOverview/AuthorizingApplicationsCard.tsx:65]:
```suggestion
          There is a deauthorization cooldown period of 45 days for tBTC and
          Random Beacon, and 6 months for TACo.
```

File: src/pages/tBTC/Bridge/Minting/ProvideData.tsx [hunk 6]
```diff
@@ -137,31 +147,60 @@ export const ProvideDataComponent: FC<{
           "The account used to generate the deposit address must be the same as the connected wallet."
         )
       }
+
+      if (!isSupportedNetwork(chainId)) {
+        throw new Error(
+          "Your wallet is on an unsupported network. Switch to a supported network"
```
Comment (derekpierre) [inline src/pages/tBTC/Bridge/Minting/ProvideData.tsx:153]:
```suggestion
          "You are currently connected to an unsupported network. Switch to a supported network"
```

**Review by derekpierre — COMMENTED** [review]
Here are my comments thus far - I'm about halfway through the changes 😅.

File: src/hooks/useFetchTvl.ts (hunk 2 above)
Comment (evandrosaturnino) [reply src/hooks/useFetchTvl.ts:79]: The `!` operator in TypeScript is a non-null assertion operator. In this case, it was used to assert that the `address` property would always be assigned a valid `string` value (indicating that a contract would always get successfully loaded). The `!` was removed because the `if (keepBonding && multicall)` check now ensures that these contracts are properly initialized. This change accommodates scenarios where `keepBonding` or `multicall` may not be set due to network unavailability, making the non-null assertion unnecessary.

File: src/hooks/ledger-live-app/useRequestEthereumAccount.ts [hunk 7]
```diff
@@ -37,20 +39,24 @@ export function useRequestEthereumAccount(): UseRequestAccountReturn {
...
     }
-  }, [account, isEmbed])
+  }, [ledgerLiveAccount, chainId, isEmbed])
 
   const requestEthereumAccount = useCallback(async () => {
     // The Goerli testnet become deprecated. However, we did not test Ledger
     // Live on Sepolia yet, so we're leaving the Goerli config for now in the
     // code.
-    const currencyId = supportedChainId === "1" ? "ethereum" : "ethereum_goerli"
+    const currencyId = isTestnetNetwork(defaultOrConnectedChainId)
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/hooks/ledger-live-app/useRequestEthereumAccount.ts:53]: Yes any of the Sepolia networks would return true in `isTestnetNetwork` function validation, this implementation follows the comment previously added about the persistance of the `ethereum_goerli` usage while Ledger Live `currencyId` hasn't been fully tested in Sepolia network.

https://github.com/threshold-network/token-dashboard/blob/63705c788b4cab1955f8b33351a059ff929d553a/src/hooks/ledger-live-app/useRequestEthereumAccount.ts#L50-L52

File: src/networks/utils/mappings.ts (hunk 4 above)
Comment (evandrosaturnino) [reply src/networks/utils/mappings.ts:34]: This function is creating mapping objects for each type of `networks` we currently have added in `networks` utils file, those mappings are being used in `utils` functions for validations. Examples below:

https://github.com/threshold-network/token-dashboard/blob/63705c788b4cab1955f8b33351a059ff929d553a/src/networks/utils/connectedNetwork.ts#L10-L31

It's not possible for the `networkType` to be undefined, because it is iterating over a hard-coded array of objects that follows strictly a certain Object type `Network[]`, previously added to assert which networks the app supports or not:

https://github.com/threshold-network/token-dashboard/blob/63705c788b4cab1955f8b33351a059ff929d553a/src/networks/utils/networks.ts#L14-L142

File: src/networks/README.md (hunk 3 above)
Comment (evandrosaturnino) [reply src/networks/README.md:126]: Thanks, this will definitely be useful for future maintenance, such as adding or removing networks that the app should support

File: src/pages/Staking/HowItWorks/StakingOverview/AuthorizingApplicationsCard.tsx (hunk 5 above)
Comment (evandrosaturnino) [reply src/pages/Staking/HowItWorks/StakingOverview/AuthorizingApplicationsCard.tsx]: Addressed in [Update src/pages/Staking/HowItWorks/StakingOverview/AuthorizingApplic…](https://github.com/threshold-network/token-dashboard/pull/785/commits/e09b33e8264584bf23f9b369848d0a04d33a129f)

File: src/pages/tBTC/Bridge/Minting/ProvideData.tsx (hunk 6 above)
Comment (evandrosaturnino) [reply src/pages/tBTC/Bridge/Minting/ProvideData.tsx]: Addressed in [pdate text in provide data validation](https://github.com/threshold-network/token-dashboard/pull/785/commits/1ea5bc9231aab45ac31c402e64af64eeb9147087)

File: src/contexts/TokenContext.tsx [hunk 8]
```diff
@@ -56,27 +56,35 @@ export const TokenContextProvider: React.FC = ({ children }) => {
     tbtcv2: tbtcv2Data,
   } = useTokenState()
 
-  const tokenContracts = [keep.contract!, nu.contract!, t.contract!]
-
-  if (featureFlags.TBTC_V2) tokenContracts.push(tbtcv2)
+  const tokenContracts = {
+    [Token.Keep]: keep.contract!,
+    [Token.Nu]: nu.contract!,
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/contexts/TokenContext.tsx:61]: I recently caught an user in Discord requesting information to how to upgrade his Nu token to T, so I believe it might still be relevant to keep those functionalities in place for the time being.

File: src/contexts/TokenContext.tsx [hunk 9]
```diff
@@ -56,27 +56,35 @@ export const TokenContextProvider: React.FC = ({ children }) => {
...
     isActive ? account! : AddressZero
   )
 
   //
   // SET T CONVERSION RATE FOR KEEP, NU
   //
-  React.useEffect(() => {
+  useEffect(() => {
     setTokenConversionRate(Token.Nu, nuConversion)
     setTokenConversionRate(Token.Keep, keepConversion)
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/contexts/TokenContext.tsx:81]: I believe they are still relevant, as I recently saw a user on Discord asking how to upgrade their Nu tokens to T. This suggests there may still be people for whom those functionalities remain important.

File: src/components/MintDurationTiers/MintDurationTiers.tsx [hunk 10]
```diff
@@ -55,9 +55,9 @@ const MintDurationTiers: FC<MintDurationTiersProps> = ({
         // is returned as is.
         const confirmations = getNumberOfConfirmationsByAmount(safeAmount)
         const durationInMinutes =
-          getDurationByNumberOfConfirmations(confirmations)
-        // Round up the minutes to the nearest half-hour
-        const hours = (Math.round(durationInMinutes / 30) * 30) / 60
+          getDurationByNumberOfConfirmations(confirmations) * 1.5
+        // Round down the minutes to the nearest half-hour
+        const hours = (Math.floor(durationInMinutes / 30) * 30) / 60
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/components/MintDurationTiers/MintDurationTiers.tsx:60]: Math.floor is used instead of Math.round to ensure that you always round down to the nearest 30-minute block avoiding overstating the number of hours.

File: src/components/MintDurationTiers/MintDurationTiers.tsx [hunk 11]
```diff
@@ -55,9 +55,9 @@ const MintDurationTiers: FC<MintDurationTiersProps> = ({
         // is returned as is.
         const confirmations = getNumberOfConfirmationsByAmount(safeAmount)
         const durationInMinutes =
-          getDurationByNumberOfConfirmations(confirmations)
-        // Round up the minutes to the nearest half-hour
-        const hours = (Math.round(durationInMinutes / 30) * 30) / 60
+          getDurationByNumberOfConfirmations(confirmations) * 1.5
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/components/MintDurationTiers/MintDurationTiers.tsx:58]: The total time required for the minting process to finalize is at least one hour, consdiering the necessary `optimisticMintingDelay` in `TBTCVault` contract time to elapse. This is not realistically aligned with how the code was originally implemented, since it was based solely on the duration of Bitcoin network confirmations.

Thus, considering that the `finalizeOptimisticMint` function must be called for the deposit to be finalized, it becomes difficult to precisely determine how long the entire process will take. On top of that, this extra waiting period can include an additional 15-30 minutes that may be required for bridging from L1 to L2 in the event of a cross-chain deposit. Consequently, increasing the total estimated time by 50% provides a more flexible estimate, taking into account both the new dynamics introduced by optimistic minting and the 15–30 minutes typically needed for L1–L2 bridging.

File: src/networks/utils/getRpcUrl.ts (hunk 1 above)
Comment (evandrosaturnino) [reply src/networks/utils/getRpcUrl.ts:8]: I tend to agree that this can be better structured based on multiple possible services, so I opened the following issue to track this https://github.com/threshold-network/token-dashboard/issues/790, and I'm going to address this in a separate PR.

File: src/contexts/TokenContext.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply src/contexts/TokenContext.tsx:61]: Ah yes, the vending machine. I was really thinking with respect to staking.

File: src/networks/utils/mappings.ts (hunk 4 above)
Comment (derekpierre) [reply src/networks/utils/mappings.ts:34]: I guess I mean that if `layer` can only be 1 or 2 and always is never undefined (?), AND `networkType` can only be `Mainnet` or `Testnet`, then you will never get to the if statement on L32. Unless I'm still misunderstanding.

Comment (evandrosaturnino) [reply src/networks/utils/mappings.ts:34]: Understood. Yes, it is reachable on every iteration because, during all of those validation checks, the function does not return and thus always proceeds to the end.

File: src/threshold-ts/mas/dapp-development-sepolia-artifacts/TacoRegistry.json [hunk 12]
```diff
@@ -0,0 +1,1352 @@
+{
+  "address": "0xE9Eb81F72cDBE9fE82f61EA527e8a3A39f5B023a",
```
Comment (derekpierre) [inline src/threshold-ts/mas/dapp-development-sepolia-artifacts/TacoRegistry.json:2]: Was this registry always hardcoded for the dashboard or are you hardcoding it now? I believe that it is typically obtained from a published npm package - https://www.npmjs.com/package/@nucypher/nucypher-contracts for all TACo domains. Same comment for other `TacoRegistry.json` files (cc @manumonti , @theref  who are more familiar)

Comment (evandrosaturnino) [reply src/threshold-ts/mas/dapp-development-sepolia-artifacts/TacoRegistry.json:2]: Addressed for mainnet TACo application artifact in [update mainnet taco application artifact source](https://github.com/threshold-network/token-dashboard/pull/785/commits/5b0d71cab0800b8f032d996e2223d63aff36dd95)

Since there is only the artifact for tapir network specifically in module, I'm going to keep the sepolia network artifact version hardcoded.

Comment (evandrosaturnino) [reply src/threshold-ts/mas/dapp-development-sepolia-artifacts/TacoRegistry.json:2]: The T token address we use on the tBTC-related Sepolia contracts is `0x46abDF5aD1726ba700794539C3dB8fE591854729`, can be checked below:

https://github.com/threshold-network/token-dashboard/blob/5b0d71cab0800b8f032d996e2223d63aff36dd95/src/web3/hooks/useT.ts#L9-L15

Which is the same address we are currently sharing in our contract addresses doc: https://docs.threshold.network/resources/contract-addresses/sepolia-testnet

From what I understood, we should we use Tapir artifact for Sepolia contracts?
If that's the case, then I will also import the Tapir artifact directly from the Nucypher module, as we did for mainnet - I will just wait for your confirmation @theref Thanks!

Comment (theref) [reply src/threshold-ts/mas/dapp-development-sepolia-artifacts/TacoRegistry.json:2]: yeah we use a different version of `T` - `0x28C35644F713c7Ee5C6A105e7AB0Fc144889a1Af` for `tapir`. I think it's still the correct choice though

Comment (evandrosaturnino) [reply src/threshold-ts/mas/dapp-development-sepolia-artifacts/TacoRegistry.json]: Got it, thank you for clarifying this.

This has been addressed in [update testnet taco registry artifacts import](https://github.com/threshold-network/token-dashboard/pull/785/commits/931da970cdc0672c5f19c728c621b8e268e9a417)

**Review by theref — DISMISSED** [review]

**Comment by Shadowfiend** [comment]
Phew, okay, sorry for the huge delay here folks, I have been underwater.

@evandrosaturnino I added all of the secrets you sent me on Keybase. Note that @derekpierre has admin access to the repo & org which means:
- You can update the secrets.
- You can add Evandro to the admin group or, if necessary, one-off Evandro as an admin for this repo specifically.

I would love to not be a blocker moving forward because obviously I'm doing a poor job of managing my time to unblock y'all.

Additional question: is it a blocker that cygnusv, Lukasz, or Piotr need to review the changes to `dashboard-ci.yml` page? If so, let's figure out who else we might add to that list who can do appropriate review and provide shorter feedback cycles.

**Review by lrsaturnino — COMMENTED** [review]
Good work. I'll submit my review shortly.

File: src/hooks/staking-applications/useBondOperatorTransaction.ts [hunk 13]
```diff
@@ -15,8 +15,7 @@ export const useBondOperatorTransaction = (
   const threshold = useThreshold()
 
   return useSendTransactionFromFn(
-    threshold.multiAppStaking[stakingAppNameToThresholdAppService[appName]]
-      .bondOperator,
+    threshold.multiAppStaking[appNameToThresholdApp[appName]]?.bondOperator!,
```
Comment (lrsaturnino) [inline src/hooks/staking-applications/useBondOperatorTransaction.ts:18]: I noticed the following expression used to pass `bondOperator` to `useSendTransactionFromFn`:

```
return useSendTransactionFromFn(
  threshold.multiAppStaking[appNameToThresholdApp[appName]]?.bondOperator!,
  onSuccess,
  onError
)
```

I understand that the optional chaining operator `(?.)` is being used to safely access the `bondOperator` property. However, the non-null assertion `(!)` that follows seems to contradict this safety check. 

Could you clarify if our expectation is that `bondOperator` is always defined once `threshold.multiAppStaking[...]` is available?

File: src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts [hunk 14]
```diff
@@ -10,27 +9,50 @@ import {
...
 
 export const useTBTCDepositDataFromLocalStorage = () => {
-  const { account } = useIsActive()
+  const { account, chainId } = useIsActive()
 
-  const [tBTCDepositData] = useLocalStorage<TBTCLocalStorageDepositData>(
-    key,
-    {}
-  )
+  const [tBTCDepositData, setTBTCLocalStorageData] =
```
Comment (lrsaturnino) [inline src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts:14]: The local storage key now incorporates `chainId` (using `${key}-${chainId?.toString()}`), which is a useful enhancement for handling data on different chains. However, I wanted to raise an inquiry regarding backward compatibility:

Previously, we were using a static key to store TBTC deposit data. With the new key format, if users have deposits stored under the old key, the app might not be able to retrieve these deposits. 

Might it be beneficial to consider a migration strategy to ensure that existing data isn’t inadvertently lost?

**Review by lrsaturnino — COMMENTED** [review]
Nice work! Here's a couple more comments about the rest of the files.

File: src/hooks/tbtc/useCheckDepositExpirationTime.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+        const { depositRevealAheadPeriod } =
+          (await threshold.tbtc.bridgeContract?.depositParameters()) ?? {}
+
+        if (!depositRevealAheadPeriod) {
+          throw new Error("Failed to retrieve deposit reveal ahead period.")
+        }
+
+        const currentTimestampInSeconds = Math.floor(Date.now() / 1000)
+        const expirationTimestamp =
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/hooks/tbtc/useCheckDepositExpirationTime.ts:24]: Good catch! This has been addressed in [fixes the reveal ahead time calculation](https://github.com/threshold-network/token-dashboard/pull/785/commits/49ad2c0374cdc5c2508a8199ad8921338652c7f6)

File: src/web3/hooks/useCheckDuplicateProviderAddress.ts [hunk 16]
```diff
@@ -17,31 +18,54 @@ const useCheckDuplicateProviderAddress = (): ((
...
+        const createdAt = keepDelegationResult?.createdAt || BigNumber.from(0)
+
+        const isProviderUsedForKeep =
+          BigNumber.isBigNumber(createdAt) && createdAt.gt(0)
+        const isProviderUsedForT = !isAddressZero(owner)
+
+        return { isProviderUsedForKeep, isProviderUsedForT }
+      } catch (error) {
+        console.error("Multicall failed:", error)
+        return {
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/web3/hooks/useCheckDuplicateProviderAddress.ts]: Addressed in [throw error when multicall is not loaded](https://github.com/threshold-network/token-dashboard/pull/785/commits/089d25b964faf787a9de2177c25084531ef2ebd9)

File: src/hooks/useIsActive.ts [hunk 17]
```diff
@@ -21,12 +24,59 @@ export const useIsActive = (): UseIsActiveResult => {
...
     deactivate: _deactivate,
+    activate: _activate,
   } = useWeb3React()
-  const { ethAccount, setEthAccount } = useLedgerLiveApp()
+  const { ethAccount, ethAccountChainId, setEthAccount } = useLedgerLiveApp()
   const ledgerLiveAppEthAddress = ethAccount?.address || undefined
+  const ledgerLiveAppEthChaindId = ethAccountChainId
   const { isEmbed } = useIsEmbed()
 
+  const switchNetwork = useCallback(async (chainId: number): Promise<void> => {
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/hooks/useIsActive.ts:36]: Good catch, addressed in: [treats error code 4901 on switching user network](https://github.com/threshold-network/token-dashboard/pull/785/commits/8a1fbd30374055dc82bfa71ebd8992449ea96dc3)

File: src/utils/tbtcLocalStorageData.ts [hunk 18]
```diff
@@ -27,17 +35,24 @@ export function write(
     [account]: newDepositData,
   }
 
-  writeStorage<TBTCLocalStorageDepositData>(key, newLocalStorageData)
+  writeStorage<TBTCLocalStorageDepositData>(
+    `${key}-${chainId?.toString()}`,
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/utils/tbtcLocalStorageData.ts:39]: This was already been ensured by the following validation
https://github.com/threshold-network/token-dashboard/blob/9e7ba9e0c721a73eb6d2a9d74444ebcfc3c7fd88/src/utils/tbtcLocalStorageData.ts#L31
Because it isn't possible for a account to be connected without chainId. However I believe it makes sense to explicitly ensure that chainId is received in order to set the local storage item, hence this has been addressed in [ensure chainid before setting data in local storage](https://github.com/threshold-network/token-dashboard/pull/785/commits/794e090398c788d2ee36666ee1ce8493348dc4cb)

File: src/threshold-ts/staking/index.ts [hunk 19]
```diff
@@ -191,51 +222,76 @@ export class Staking implements IStaking {
...
-        args: [stakingProvider],
-      },
-      {
-        interface: this._staking.interface,
-        address: this._staking.address,
-        method: "stakes",
-        args: [stakingProvider],
-      },
-      {
+    const multicalls: ContractCall[] = []
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/threshold-ts/staking/index.ts:227]: Good catch! Addressed in [returns empty stake when no staking contract is loaded](https://github.com/threshold-network/token-dashboard/pull/785/commits/77c26efb6bc7864d55732769c01419aa201a0904)

File: src/threshold-ts/tbtc/sepolia-artifacts/BaseL1BitcoinDepositor.json
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/threshold-ts/tbtc/sepolia-artifacts/BaseL1BitcoinDepositor.json:1]: Addressed in [updates the l1 bitcoin depositor artifact file for base](https://github.com/threshold-network/token-dashboard/pull/785/commits/a64f6758868d525223f8d1de18dc7d31078f709d)

File: src/store/staking-applications/effects.ts [hunk 20]
```diff
@@ -50,31 +65,33 @@ export const getSupportedAppsEffect = async (
       await listenerApi.extra.threshold.multiAppStaking.getSupportedAppsAuthParameters()
     const payload = {
       tbtc: {
-        minimumAuthorization: data.tbtc.minimumAuthorization.toString(),
+        minimumAuthorization: data.tbtc!.minimumAuthorization.toString(),
 
         authorizationDecreaseDelay:
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/store/staking-applications/effects.ts:70]: The non-null assertion can be maintained here, because we are ensuring that all the three services were loaded correctly before proceeding. The services are ecdsa (tbtc), random beacon and taco apps:

https://github.com/threshold-network/token-dashboard/blob/9e7ba9e0c721a73eb6d2a9d74444ebcfc3c7fd88/src/store/staking-applications/effects.ts#L41-L45

The auth parameters are gotten based on the availability of these services in the `getSupportedAppsAuthParameters()` method:

https://github.com/threshold-network/token-dashboard/blob/9e7ba9e0c721a73eb6d2a9d74444ebcfc3c7fd88/src/threshold-ts/mas/index.ts#L83-L108

File: src/hooks/tbtc/useFetchDepositDetails.ts [hunk 21]
```diff
@@ -44,12 +62,19 @@ export const useFetchDepositDetails = (depositKey: string | undefined) => {
 
         if (!deposit) {
           throw new Error(
-            "Could not find `DepositRevealed` event by given deposit key."
+            "Could not find DepositRevealed event by given deposit key."
           )
         }
 
+        const l1BitcoinDepositorDepositStatus = !!threshold.tbtc
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply src/hooks/tbtc/useFetchDepositDetails.ts:69]: Thanks for the feedback! I’d like to clarify that the !! is applied to threshold.tbtc.l1BitcoinDepositorContract to check its existence. 

If it exists, we call deposits(depositKey) to return the status from the L1BitcoinDepositor contract; otherwise, we return DepositState.UNKNOWN. So, the deposit status isn’t being coerced to a boolean in this line, and the current approach correctly distinguishes the two cases.

File: src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts (hunk 14 above)
Comment (evandrosaturnino) [reply src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts:14]: When a user connects their wallet to the app, all of their deposit iterations are visible in the History tab on the mint page. Additionally, users can resume a deposit using the generated JSON file. 
 
The local storage feature mainly provides the convenience of automatically loading data. For this migration purposes, I believe that allowing users to access their deposits through the History tab is sufficient. Meanwhile, new deposits, loaded based on the selected network, will be automatically saved to local storage, ensuring a seamless experience upon connection.

Comment (lrsaturnino) [reply src/hooks/tbtc/useTBTCDepositDataFromLocalStorage.ts:14]: thanks @evandrosaturnino. I agree.

**Review by lrsaturnino — APPROVED** [review]
LGTM

**Review by lrsaturnino — APPROVED** [review]
Amazing work. LGTM!
